### DISCUSSION TITLE
Is there a working script to reproduce results of enumeration of cages in your papers?

### DISCUSSION THREAD

**Discussion by ridakbr** [lukasturcani/stk#303] [Q&A]
I want to reproduce the results in Figure 7 and 11 from "An evolutionary algorithm for the discovery of porous organic cages" and Figure 5 from "High-throughput discovery of organic cages and catenanes using computational screening fused with robotic synthesis", but I am unable at all! and it is like I have been trying to reproduce it for 1 week.

For the preinitial step, I got no problem installing stk codes and its prerequisites like I can produce the polymer.mol, view it in my ChemDraw, and reproduce the figure like in Docs (copy paste the code from https://stk.readthedocs.io/en/latest/construction_overview.html, put the code in text editor, and run it in Terminal).

Then, I tried to do the same for the code in https://stk.readthedocs.io/en/latest/stk.ea.mutation.mutators.molecule.random_topology_graph.html and Voila! This is what I got:

File "yourmutation.py", line 14
   replacement_funcs = (
   ^
SynthaxError: invalid syntax

So, where is the full real script for the cage enumeration? Your example script in Docs doesn’t give any insight to reproduce the results in those papers for broad audiences.

**Comment by lukasturcani** [comment]
Hi there, For "High-throughput discovery of organic cages and catenanes using computational screening fused with robotic synthesis" I'm not an author on that paper, so I'm afraid the support I can give you is limited, I suggest you get in touch with the corresponding authors of that work. I will also note that the paper likely used an older version of stk, so there is a decent chance that whatever script was used to produce the results of that work would not work with the lastest version of stk.

However, I am happy to give you support as far as I can. In https://stk.readthedocs.io/en/latest/stk.ea.mutation.mutators.molecule.random_topology_graph.html, there is  typo in the docs,  `stk.cage.FourPlusSix((bb1, bb2)` should have been `stk.cage.FourPlusSix((bb1, bb2))` I've opened an issue to track this, https://github.com/lukasturcani/stk/issues/304 )

Try the following code (EDIT: I've edited this so that the snippet actually works, the original snippet I posted had inappropriate aldehyde smiles)

```python

import stk

# Create a molecule which is to be mutated.
bb1 = stk.BuildingBlock('NCCN', [stk.PrimaryAminoFactory()])
bb2 = stk.BuildingBlock(
    smiles='O=CCC(C=O)CC=O',
    functional_groups=[stk.AldehydeFactory()],
)
cage = stk.MoleculeRecord(
    topology_graph=stk.cage.FourPlusSix((bb1, bb2)),
)

# Create functions which replace the topology graph.
replacement_funcs = (
    lambda graph:
        stk.cage.TwoPlusThree(graph.get_building_blocks()),
    lambda graph:
        stk.cage.EightPlusTwelve(graph.get_building_blocks()),
    lambda graph:
        stk.cage.TwentyPlusThirty(graph.get_building_blocks()),
)

# Create the mutator.
random_topology = stk.RandomTopologyGraph(
    replacement_funcs=replacement_funcs,
    random_seed=21,
)

# Mutate a molecule.
mutation_record1 = random_topology.mutate(cage)

# Mutate the molecule a second time.
mutation_record2 = random_topology.mutate(cage)
```

**Reply by lukasturcani** [reply]
For "An evolutionary algorithm for porous organic cages", this definitely used what is an old version stk, so if you want to use the script that was used to generate those figures you would have to checkout an old commit of stk and use an appropriate input file. Don't have the input files used for the paper on me unfortunately. In any case, if you would like to construct the equivalent EA using the current version of stk, I'm happy to help you with any issues you run into, this is the way I would go about reproducing the results. Note that the paper uses proprietary software for molecular optimization (Macromodel) and a proprietary database of precursors. For more specific questions about the paper I reccomend getting in touch with corresponding authors.

**Reply by ridakbr** [reply]
Thank you for your quick response :)

Ah, I was expecting more on stk so that stk can make different topologies like a graph emulator. Please keep working hard on the project because I need this.

By the way, I have tried your code described above and it still gave me error like this.
Traceback (most recent call last):
  File "yourmutation.py", line 10, in <module>
    topology_graph=stk.cage.FourPlusSix((bb1, bb2)),
  File "/home/labbioinfo/.local/lib/python3.8/site-packages/stk/molecular/topology_graphs/cage/cage.py", line 613, in __init__
    building_block_vertices = self._normalize_building_blocks(
  File "/home/labbioinfo/.local/lib/python3.8/site-packages/stk/molecular/topology_graphs/cage/cage.py", line 665, in _normalize_building_blocks
    return cls._get_building_block_vertices(
  File "/home/labbioinfo/.local/lib/python3.8/site-packages/stk/molecular/topology_graphs/cage/cage.py", line 821, in _get_building_block_vertices
    raise ValueError(
ValueError: If there are multiple building blocks with the same number of functional groups, building_block_vertices must be set explicitly.

I guess the code needs more complex structure of the molecules for the building blocks but seems the problem is multifolds.

**Reply by lukasturcani** [reply]
Arrgh, really sorry that didn't work either. I fucked up the smiles on the aldehyde. This should work, as I've actually run it myself

```python
import stk

# Create a molecule which is to be mutated.
bb1 = stk.BuildingBlock('NCCN', [stk.PrimaryAminoFactory()])
bb2 = stk.BuildingBlock(
    smiles='O=CCC(C=O)CC=O',
    functional_groups=[stk.AldehydeFactory()],
)
cage = stk.MoleculeRecord(
    topology_graph=stk.cage.FourPlusSix((bb1, bb2)),
)
stk.MolWriter().write(
    molecule=cage.get_molecule(),
    path='original.mol',
)

# Create functions which replace the topology graph.
replacement_funcs = (
    lambda graph:
        stk.cage.TwoPlusThree(graph.get_building_blocks()),
    lambda graph:
        stk.cage.EightPlusTwelve(graph.get_building_blocks()),
    lambda graph:
        stk.cage.TwentyPlusThirty(graph.get_building_blocks()),
)

# Create the mutator.
random_topology = stk.RandomTopologyGraph(
    replacement_funcs=replacement_funcs,
    random_seed=21,
)

# Mutate a molecule.
mutation_record1 = random_topology.mutate(cage)
stk.MolWriter().write(
    molecule=mutation_record1.get_molecule_record().get_molecule(),
    path='mutant1.mol',
)

# Mutate the molecule a second time.
mutation_record2 = random_topology.mutate(cage)
stk.MolWriter().write(
    molecule=mutation_record2.get_molecule_record().get_molecule(),
    path='mutant2.mol',
)

```

**Reply by lukasturcani** [reply]
Yeah I'm doing my best to fit time into the project but unfortunately it's very much something I'm only able to work on in my free time.  Atm the `RandomTopologyGraph` will only substitute different pre-existing topology graphs. But I'm working on a feature here: https://github.com/lukasturcani/stk/issues/206 , which should allow actual modification of the underlying molecular graph as part of the EA.

**Reply by ridakbr** [reply]
Thank you so much! Now the code is working as what I expected. I have a small question.

Is it possible to input number of molecules like stk.BuildingBlock('100 molecules in one file') rather than inputing multiple SMILES manually? I have a ligand database, functionalized with desired chemical functional groups.

I am surely going to ask the first author of the paper mentioning discovery of catenane cages. I have checked its Supplementary Materials but seems the authors on the paper forgot to put the program in computational methods.

**Reply by lukasturcani** [reply]
Assuming that the ligands are in the file written in SMILES, with each SMILES on it own line, I would write something like this

```python

with open("myligandfile") as ligand_file:
    building_blocks = tuple(
        stk.BuildingBlock(ligand_smiles, [stk.SomeFunctionalGroupFactory()]) 
        for lingand_smiles in ligand_file 
    )

```

where  you change `stk.SomeFunctionalGroupFactory()` to whatever is needed for your ligand

**Reply by ridakbr** [reply]
I'm thinking if the mutator may be able to perform n-mer of each building blocks in same topology (like assemblying 2, 3, so on molecules of each building blocks in same (origins) of topology). I think this feature would unearth more structural complexity on same topology.

**Reply by ridakbr** [reply]
> Yeah I'm doing my best to fit time into the project but unfortunately it's very much something I'm only able to work on in my free time. Atm the `RandomTopologyGraph` will only substitute different pre-existing topology graphs. But I'm working on a feature here: #206 , which should allow actual modification of the underlying molecular graph as part of the EA.

This would be very great! To capture topological mutation like from [4+6] into [n1+n2] or [n1+n2+...n(k)] without having to specify each line in replacement_funcs.

**Reply by lukasturcani** [reply]
hmm, I'm not sure I follow, could you give an example of inputs and desired outputs?

**Reply by ridakbr** [reply]
> I'm thinking if the mutator may be able to perform n-mer of each building blocks in same topology (like assemblying 2, 3, so on molecules of each building blocks in same (origins) of topology). I think this feature would unearth more structural complexity on same topology.

I mean like Figure 3a in "High-throughput discovery of organic cages and catenanes using computational screening fused with robotic synthesis". Here the authors make the multimer of the building blocks (fused, spiro, linkered with other molecule) before making into cage.

I am sure these authors use kind of molecular assembler but I think my idea is out of the scope of stk itself.

**Reply by ridakbr** [reply]
> Assuming that the ligands are in the file written in SMILES, with each SMILES on it own line, I would write something like this
> 
> ```python
> with open("myligandfile") as ligand_file:
>     building_blocks = tuple(
>         stk.BuildingBlock(ligand_smiles, [stk.SomeFunctionalGroupFactory()]) 
>         for lingand_smiles in ligand_file 
>     )
> ```
> 
> where you change `stk.SomeFunctionalGroupFactory()` to whatever is needed for your ligand

By the way, which line I should put the code on, from your worked aldehyde snippet. I want to test it directly.

**Reply by lukasturcani** [reply]
> > Yeah I'm doing my best to fit time into the project but unfortunately it's very much something I'm only able to work on in my free time. Atm the `RandomTopologyGraph` will only substitute different pre-existing topology graphs. But I'm working on a feature here: #206 , which should allow actual modification of the underlying molecular graph as part of the EA.
> 
> This would be very great! To capture topological mutation like from [4+6] into [n1+n2] or [n1+n2+...n(k)] without having to specify each line in replacement_funcs.

I think this would mostly be used when evolving molecules along the lines of a graph-based EA (https://pubs.acs.org/doi/10.1021/ci034290p). I don't think it would replace the use-case of `replacement_funcs`. The issue with modifying [4+6] to [n1+n2] is that if you remove a node or edge from a [4+6] topology you do not have a guarantee that you still end up with a cage topology. The purpose of `replacement_funcs` is that you can guarantee you always have a valid cage topology after a topolgy graph mutation + you can guarantee which cage topology graphs are possible. I don't have a better idea of how else to provide this guarantee :(. However if you do not care about this guarantee then yeah it would be possible to write a mutator which changes the topology graph. But I expect that very soon your population would end up with no cage topology graphs.

**Reply by lukasturcani** [reply]
> > Assuming that the ligands are in the file written in SMILES, with each SMILES on it own line, I would write something like this
> > ```python
> > with open("myligandfile") as ligand_file:
> >     building_blocks = tuple(
> >         stk.BuildingBlock(ligand_smiles, [stk.SomeFunctionalGroupFactory()]) 
> >         for lingand_smiles in ligand_file 
> >     )
> > ```
> > 
> > 
> > where you change `stk.SomeFunctionalGroupFactory()` to whatever is needed for your ligand
> 
> By the way, which line I should put the code on, from your worked aldehyde snippet. I want to test it directly.

This shoud work as a stand-alone script. You can put the snippet anywhere you like really. You will just have to replace `"myligandfile"` with whatever the path to your file is and `stk.SomeFunctionalGroupFactory()` with whatever functional group factory you want to use.

**Reply by ridakbr** [reply]
> > > Yeah I'm doing my best to fit time into the project but unfortunately it's very much something I'm only able to work on in my free time. Atm the `RandomTopologyGraph` will only substitute different pre-existing topology graphs. But I'm working on a feature here: #206 , which should allow actual modification of the underlying molecular graph as part of the EA.
> > 
> > 
> > This would be very great! To capture topological mutation like from [4+6] into [n1+n2] or [n1+n2+...n(k)] without having to specify each line in replacement_funcs.
> 
> I think this would mostly be used when evolving molecules along the lines of a graph-based EA (https://pubs.acs.org/doi/10.1021/ci034290p). I don't think it would replace the use-case of `replacement_funcs`. The issue with modifying [4+6] to [n1+n2] is that if you remove a node or edge from a [4+6] topology you do not have a guarantee that you still end up with a cage topology. The purpose of `replacement_funcs` is that you can guarantee you always have a valid cage topology after a topolgy graph mutation + you can guarantee which cage topology graphs are possible. I don't have a better idea of how else to provide this guarantee :(. However if you do not care about this guarantee then yeah it would be possible to write a mutator which changes the topology graph. But I expect that very soon your population would end up with no cage topology graphs.

The author of the paper you mentioned, Ben McKay, is the developer behind GENG, a graph-based enumeration tool and the tool was used thoroughly in making ligand database with focus on geometrical novelty of 3D molecules, the GDB-17 database. I was thinking perhaps you can implement GENG code along your code to guarantee the specification of the graphs.

**Reply by lukasturcani** [reply]
> > I'm thinking if the mutator may be able to perform n-mer of each building blocks in same topology (like assemblying 2, 3, so on molecules of each building blocks in same (origins) of topology). I think this feature would unearth more structural complexity on same topology.
> 
> I mean like Figure 3a in "High-throughput discovery of organic cages and catenanes using computational screening fused with robotic synthesis". Here the authors make the multimer of the building blocks (fused, spiro, linkered with other molecule) before making into cage.
> 
> I am sure these authors use kind of molecular assembler but I think my idea is out of the scope of stk itself.

Is Figure 3a not just showing a standard cage construction where you combine two building blocks + a topology graph to make a cage?

**Reply by ridakbr** [reply]
I will be back again on there for putting the ligand database input file in the code :)

**Reply by lukasturcani** [reply]
> > > > Yeah I'm doing my best to fit time into the project but unfortunately it's very much something I'm only able to work on in my free time. Atm the `RandomTopologyGraph` will only substitute different pre-existing topology graphs. But I'm working on a feature here: #206 , which should allow actual modification of the underlying molecular graph as part of the EA.
> > > 
> > > 
> > > This would be very great! To capture topological mutation like from [4+6] into [n1+n2] or [n1+n2+...n(k)] without having to specify each line in replacement_funcs.
> > 
> > 
> > I think this would mostly be used when evolving molecules along the lines of a graph-based EA (https://pubs.acs.org/doi/10.1021/ci034290p). I don't think it would replace the use-case of `replacement_funcs`. The issue with modifying [4+6] to [n1+n2] is that if you remove a node or edge from a [4+6] topology you do not have a guarantee that you still end up with a cage topology. The purpose of `replacement_funcs` is that you can guarantee you always have a valid cage topology after a topolgy graph mutation + you can guarantee which cage topology graphs are possible. I don't have a better idea of how else to provide this guarantee :(. However if you do not care about this guarantee then yeah it would be possible to write a mutator which changes the topology graph. But I expect that very soon your population would end up with no cage topology graphs.
> 
> The author of the paper you mentioned, Ben McKay, is the developer behind GENG, a graph-based enumeration tool and the tool was used thoroughly in making ligand database with focus on geometrical novelty of 3D molecules, the GDB-17 database. I was thinking perhaps you can implement GENG code along your code to guarantee the specification of the graphs.

Oh nice, sounds interesting, I will check it out

**Reply by ridakbr** [reply]
> > > I'm thinking if the mutator may be able to perform n-mer of each building blocks in same topology (like assemblying 2, 3, so on molecules of each building blocks in same (origins) of topology). I think this feature would unearth more structural complexity on same topology.
> > 
> > 
> > I mean like Figure 3a in "High-throughput discovery of organic cages and catenanes using computational screening fused with robotic synthesis". Here the authors make the multimer of the building blocks (fused, spiro, linkered with other molecule) before making into cage.
> > I am sure these authors use kind of molecular assembler but I think my idea is out of the scope of stk itself.
> 
> Is Figure 3a not just showing a standard cage construction where you combine two building blocks + a topology graph to make a cage?

You should check the middle part of Figure 3a, the molecules numbered from 1 to 10 (those 10 molecules belong on same topology). They diversify the building blocks too before making the cage :)

**Reply by lukasturcani** [reply]
> > > > I'm thinking if the mutator may be able to perform n-mer of each building blocks in same topology (like assemblying 2, 3, so on molecules of each building blocks in same (origins) of topology). I think this feature would unearth more structural complexity on same topology.
> > > 
> > > 
> > > I mean like Figure 3a in "High-throughput discovery of organic cages and catenanes using computational screening fused with robotic synthesis". Here the authors make the multimer of the building blocks (fused, spiro, linkered with other molecule) before making into cage.
> > > I am sure these authors use kind of molecular assembler but I think my idea is out of the scope of stk itself.
> > 
> > 
> > Is Figure 3a not just showing a standard cage construction where you combine two building blocks + a topology graph to make a cage?
> 
> You should check the middle part of Figure 3a, the molecules numbered from 1 to 10 (those 10 molecules belong on same topology). They diversify the building blocks too before making the cage :)

Not sure if I'm getting this right but as far as I can tell, those building blocks were used to make a cage of the same topology, but only a single one is used at a time. so for Fig 3a, for building blocks 1-10, there are 30 cages, each consists of two buildings blocks, one building block from A, B and C on the left of the figure, and one from building blocks 1 to 10 in the middle of the figure.

Not sure what you mean by diversifying.

**Reply by ridakbr** [reply]
> > > > > I'm thinking if the mutator may be able to perform n-mer of each building blocks in same topology (like assemblying 2, 3, so on molecules of each building blocks in same (origins) of topology). I think this feature would unearth more structural complexity on same topology.
> > > > 
> > > > 
> > > > I mean like Figure 3a in "High-throughput discovery of organic cages and catenanes using computational screening fused with robotic synthesis". Here the authors make the multimer of the building blocks (fused, spiro, linkered with other molecule) before making into cage.
> > > > I am sure these authors use kind of molecular assembler but I think my idea is out of the scope of stk itself.
> > > 
> > > 
> > > Is Figure 3a not just showing a standard cage construction where you combine two building blocks + a topology graph to make a cage?
> > 
> > 
> > You should check the middle part of Figure 3a, the molecules numbered from 1 to 10 (those 10 molecules belong on same topology). They diversify the building blocks too before making the cage :)
> 
> Not sure if I'm getting this right but as far as I can tell, those building blocks were used to make a cage of the same topology, but only a single one is used at a time. so for Fig 3a, for building blocks 1-10, there are 30 cages, each consists of two buildings blocks, one building block from A, B and C on the left of the figure, and one from building blocks 1 to 10 in the middle of the figure.
> 
> Not sure what you mean by diversifying.

Oh sorry, I forgot to check that they already had two different libraries for coupling these aldehydes libraries and imines libraries to make the cages. I thought their imine libraries were a functionalized form of that aldehyde libraries and so I thought they diversified from the aldehydes to imines.
